New Rule: requireDictionaryWords

jacksonrayhamilton opened the pull request:
For #1199; implements spellcheck for identifiers.

Offers fine-grained control over the (sub)words and names that can be used in / for identifiers and properties. The user can restrictively allow words, such that he can operate with 3rd-party APIs but minimize the pollution they might have on his dictionary. The user can also exclude words that he doesn't like. (For examples, see the docblock in `require-dictionary-words.js`.)

Not sure what the general stance is on including English dictionaries. I think it would be convenient to bundle them, but if you want we can externalize them as optional npm packages.

`js-names.json` and `js-property-names.json` are sets of own properties found on the global objects and the global objects' `prototype` objects in Node 0.10.x. I'm not sure how expansive or configurable those lists should be, but I expect some combination or intersection of ES5, ES6, browser and server environments would suffice.

I'm also not sure if there are "too many" tests. I like to be thorough when developing new features, but I can reduce them if you want.

TheSavior commented:
FWIW, I'm against having these dictionary lists in JSCS. If we are really going to have this rule, I don't think we should have to version bump JSCS every time some new property gets added to a generic javascript object. 

Also, if we are really going to have this rule, I don't think we should error every time an identifier is found that doesn't match a dictionary. I would think it should only be on definitions and assignments.

jacksonrayhamilton commented:
Thanks for the feedback Eli.

I updated the branch to only error on definitions and assignments; that reduced the complexity quite a bit. I don't need those built-ins lists anymore because those objects and properties are always being accessed, and rarely if ever declared. Not having to blacklist as many accessed names allowed me to remove 19 of the formerly-55 additional dictionary words I had been using for a decent-sized (~100 file) test project.

I also externalized the wordlists. The first (and only) one can be installed with [`npm install wordlist-english`](https://www.npmjs.com/package/wordlist-english); if it isn't installed and you try to use spellcheck, JSCS will throw an informative error telling you to install it. Anyone can add a new wordlist by creating his own `wordlist-*` package, be it published or `npm link`'d.

I hope we can scale this feature to a manageable and befitting size for JSCS. It looks like externalization has greatly simplified it. I really like the high level of granularity currently afforded, but the feature could be simplified further by removing some of that customization. But then again, based on the test coverage it looks like most of that work is already done anyway.

I expect the majority of maintenance associated with this rule might come with:

- comment / docblock support
- non-Latin language word splitting

if either of those are desired.

The quality of the wordlists (which is likely to be the most important issue for users) can be managed independently of JSCS.

I do think JSCS is a good place for this; I need a command-line utility that points to exactly the line and column where I need to fix a stylistic mistake. Ensuring good naming conventions is something that I do at the same time as when I validate the spaces around my braces.

TheSavior commented on lib/checker.js:
```diff
@@ -1,3 +1,6 @@
+var codependency = require('codependency');
```
These lines could be defined in the specific rule that uses it, right?

TheSavior commented on lib/rules/require-dictionary-words.js:
```diff
@@ -0,0 +1,331 @@
+ *
+ * Type: `Boolean` or `Object`
+ *
+ * Values:
```
This doc should probably describe how to install third party word lists, correct?

TheSavior commented on test/specs/rules/require-dictionary-words.js:
```diff
@@ -0,0 +1,204 @@
+var Checker = require('../../../lib/checker');
```
You commented that there was a nice error message for people trying to use a word list that wasn't installed. There should probably be a test that verifies that.

jacksonrayhamilton replied:
Added at the bottom.

TheSavior commented on test/specs/rules/require-dictionary-words.js:
```diff
@@ -0,0 +1,204 @@
+            checker.configure({ requireDictionaryWords: true, esnext: true });
+        });
+
+        it('should report non-words in ES6', function() {
```
Could use `it('should not report real words in es6');`

TheSavior commented:
These changes are much better. I don't think we should define any peer dependency by default for this rule. I'd be happier if it just gave a message when the user tried to use it and needed to install a word list.

@markelog Someone else on the project should probably weigh in here with their thoughts. 

Now that everything is externalized I don't have nearly the same level of problem I had with it previously, however I'm still not totally happy that it requires us adding an additional `npm install` command to our test suite.

jacksonrayhamilton commented:
Removed the `codependency` package and just used `require`. Improved the error messages. Coverage is now at 100% for this rule.

Added `wordlist-english` as a regular dependency; maintainers and CI systems won't need to specially install it, and neither will [most](http://en.wikipedia.org/wiki/Non-English-based_programming_languages#Prevalence_of_English-based_programming_languages) end-users. (Sorry to people from other countries who have to put up with this language.)

TheSavior commented:
Hmm, perhaps I wasn't communicating well what I was thinking. 

I don't think we should list any word list in any dependency. I think this rule should try to load the request word list, and if it doesn't exist then it prints a message saying that the user needs to install a word list.

jacksonrayhamilton commented:
Okay. I guess for the tests we can mock an npm package by copying it into `node_modules` in `beforeEach` and deleting it in `afterEach`.

IMO it would be kind of annoying for all our end users to have to add a special hook to their continuous integration scripts to `cd` and install a wordlist. I guess that is already theoretically true of other languages if anyone ever creates another `wordlist-*` package, but I'd like to think we cover the most common case by including this one.

The only advantages I see in *not* including an English wordlist are 1) reducing bias (though this particular language, and apparently the programming community, are already biased towards English, so we would be kind of ignoring reality), and 2) reducing `npm install` time. But you don't run `npm install` very often, and an occasionally-uncached ~1.5 MB isn't going to make you miss a deadline.

zxqfox commented:
Why not to have this as a plugin? It's a really huge work which would be hard to support without a deep knowledge of internal logic. Also it will give us some advantages: clean issue list for this functionality, shorter release cycle, separate team which know more about different languages and less about jscs itself, etc.

Also I think that it would be great to have this plugin in `jscs-dev` group. @mikesherov @mdevils @markelog what you think guys?

mikesherov commented:
I too would love this as a plugin that is maintained under the JSCS organization. @jacksonrayhamilton, thoughts?

jacksonrayhamilton commented:
Oh look, there's a plugin API. Sure.

jacksonrayhamilton commented:
Created [a repo](https://github.com/jacksonrayhamilton/jscs-spellcheck). Can transfer ownership.

I couldn't get coverage working; `ReferenceError: __unitCoverage__ is not defined`. If anyone could help with that, that'd be great.

Also not sure how many GitHub services will be afforded to it; can remove those from the README, etc.

mikesherov commented:
@jacksonrayhamilton that would be awesome, and then I can add you back as a maintainer to the repo @mdevils can you point @jacksonrayhamilton in the right direction with getting unit coverage working?

jacksonrayhamilton commented:
Okay. Just let me know [when](https://help.github.com/articles/transferring-a-repository/#transferring-from-a-user-to-an-organization) and I'll send it over.

zxqfox commented:
Just a link to Plugin API: https://github.com/jscs-dev/node-jscs/wiki/Plugin-API
Feedback would be appreciated ;-)

markelog commented:
Niiice!

TheSavior commented:
@zxqfox, Is there a discussion somewhere about the plugin api? I hadn't seen it until you linked it, but my concern is that it doesn't enable people to have a node_module that is there style guide that might contain more than just jscs but also jshint, editor_config, etc.

zxqfox commented:
@TheSavior #739 this only. Also it's related to many other things like how it install them #863 (because, jscs should be installed globally for IDEs, but plugins can be local). That's why we thought about `jscs-cli` or something else #894. Etc. Perhaps it's a time to collect all suggestion to the one issue to rule them all.

I've thought a lot of something to rule jshint+editor_config+etc., but the problem here is software that expect files in project root. And I can suggest here only copying or symlinking (if you don't care about win) files from `node_modules/ultimate-cs/configs/*` on postinstall/postupdate.

zxqfox commented:
@mikesherov @markelog @mdevils Folks, we need your wish here (and access rights) to transfer https://github.com/jacksonrayhamilton/jscs-spellcheck to `jscs-dev` group.

markelog commented:
@jacksonrayhamilton repository is created and invite has been sent!

jacksonrayhamilton commented:
Thanks @markelog! Just committed.

I would still like to get all the CI goodies (testing, coverage) set up. Then we can publish `v1.0.0`.

markelog commented:
Just give admin rights, let me know if you need something else

jacksonrayhamilton commented:
Still getting that coverage error and frankly have no idea how to fix it:

```bash
~/projects/jscs-spellcheck (master) λ npm run coverage
> jscs-spellcheck@1.0.0 coverage /home/jackson/projects/jscs-spellcheck
> unit-coverage run -p common


/home/jackson/projects/jscs-spellcheck/test/specs/rules/require-dictionary-words.js:375
    before(__unitCoverage__.initialize);
           ^
ReferenceError: __unitCoverage__ is not defined
    at Object.<anonymous> (/home/jackson/projects/jscs-spellcheck/test/specs/rules/require-dictionary-words.js:375:12)
    at Module._compile (module.js:456:26)
    at Object.require.extensions..js (/home/jackson/projects/jscs-spellcheck/node_modules/unit-coverage/lib/require-replacement.js:45:23)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at /home/jackson/projects/jscs-spellcheck/node_modules/mocha/lib/mocha.js:192:27
    at Array.forEach (native)
    at Mocha.loadFiles (/home/jackson/projects/jscs-spellcheck/node_modules/mocha/lib/mocha.js:189:14)
    at Mocha.run (/home/jackson/projects/jscs-spellcheck/node_modules/mocha/lib/mocha.js:422:31)
    at Object.<anonymous> (/home/jackson/projects/jscs-spellcheck/node_modules/mocha/bin/_mocha:398:16)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:935:3

/home/jackson/projects/jscs-spellcheck/node_modules/unit-coverage/node_modules/vow/lib/vow.js:92
            throw e;
                  ^
Error: Command failed: node_modules/.bin/mocha --compilers js:/home/jackson/projects/jscs-spellcheck/node_modules/unit-coverage/lib/require-replacement.js

/home/jackson/projects/jscs-spellcheck/test/specs/rules/require-dictionary-words.js:375
    before(__unitCoverage__.initialize);
           ^
ReferenceError: __unitCoverage__ is not defined
    at Object.<anonymous> (/home/jackson/projects/jscs-spellcheck/test/specs/rules/require-dictionary-words.js:375:12)
    at Module._compile (module.js:456:26)
    at Object.require.extensions..js (/home/jackson/projects/jscs-spellcheck/node_modules/unit-coverage/lib/require-replacement.js:45:23)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at /home/jackson/projects/jscs-spellcheck/node_modules/mocha/lib/mocha.js:192:27
    at Array.forEach (native)
    at Mocha.loadFiles (/home/jackson/projects/jscs-spellcheck/node_modules/mocha/lib/mocha.js:189:14)
    at Mocha.run (/home/jackson/projects/jscs-spellcheck/node_modules/mocha/lib/mocha.js:422:31)
    at Object.<anonymous> (/home/jackson/projects/jscs-spellcheck/node_modules/mocha/bin/_mocha:398:16)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:935:3

    at ChildProcess.<anonymous> (/home/jackson/projects/jscs-spellcheck/node_modules/unit-coverage/lib/utils/subprocess.js:24:26)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:766:16)
    at Process.ChildProcess._handle.onexit (child_process.js:833:5)

npm ERR! jscs-spellcheck@1.0.0 coverage: `unit-coverage run -p common`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the jscs-spellcheck@1.0.0 coverage script.
npm ERR! This is most likely a problem with the jscs-spellcheck package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     unit-coverage run -p common
npm ERR! You can get their info via:
npm ERR!     npm owner ls jscs-spellcheck
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 3.13.0-24-generic
npm ERR! command "/home/jackson/.nvm/v0.10.37/bin/node" "/home/jackson/.nvm/v0.10.37/bin/npm" "run" "coverage"
npm ERR! cwd /home/jackson/projects/jscs-spellcheck
npm ERR! node -v v0.10.37
npm ERR! npm -v 1.4.28
npm ERR! code ELIFECYCLE
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/jackson/projects/jscs-spellcheck/npm-debug.log
npm ERR! not ok code 0
```

Tried to parallel JSCS's setup as closely as possible but for some reason it just won't work. Help please?

mdevils commented:
I will take a look soon

markelog commented:
@jacksonrayhamilton other then that, can we say this plugin is tested and ready to be used?

jacksonrayhamilton commented:
Yep, I believe coverage was around 100%, and the plugin worked when used in a personal project.

Long term I might want to make it possible to use top-level devDependency wordlists rather than having to manually npm install them in the jscs-spellcheck directory. I'd also like to make some programming acronym and programming-ism wordlists. But I don't think those issues are barring a release.

mdevils commented:
@jacksonrayhamilton Sorry for a delay.

I fixed the issue with code coverage: https://github.com/jscs-dev/jscs-spellcheck/commit/382c11456457cb827d48e8cd62676ba82200d904

I will try to fix this in `unit-coverage` also later.

jacksonrayhamilton commented:
Thank you very much @mdevils! Back to 100% now.

jacksonrayhamilton commented:
v1.0.0 is published. Thanks again to the organization for its help.

mikesherov commented:
Wooo! Thanks again @jacksonrayhamilton
